### ISSUE TITLE
Bundle is not compatible with Symfony Flex

### ISSUE THREAD

**Issue by mdogancay** [azine/email-bundle#56]
```
composer req azine/email-bundle dev-master
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Restricting packages listed in "symfony/symfony" to "4.2.*"
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for __root__ dev-master -> satisfiable by __root__[dev-master].
    - azine/email-bundle dev-master requires azine/emailupdateconfirmation-bundle ~1.0 -> satisfiable by azine/emailupdateconfirmation-bundle[1.0.0, 1.0.1].
    - azine/emailupdateconfirmation-bundle 1.0.0 requires symfony/symfony ~2.7|^3.0|^4.0 -> satisfiable by symfony/symfony[v4.2.0, v4.2.1, v4.2.2, v4.2.3, v4.2.4].
    - azine/emailupdateconfirmation-bundle 1.0.1 requires symfony/symfony ~2.7|^3.0|^4.0 -> satisfiable by symfony/symfony[v4.2.0, v4.2.1, v4.2.2, v4.2.3, v4.2.4].
    - symfony/symfony v4.2.0 conflicts with __root__[dev-master].
    - symfony/symfony v4.2.1 conflicts with __root__[dev-master].
    - symfony/symfony v4.2.2 conflicts with __root__[dev-master].
    - symfony/symfony v4.2.3 conflicts with __root__[dev-master].
    - symfony/symfony v4.2.4 conflicts with __root__[dev-master].
    - Installation request for azine/email-bundle dev-master -> satisfiable by azine/email-bundle[dev-master].


Installation failed, reverting ./composer.json to its original content.
```

**Comment by azine** [comment]
looking at the output, it more looks like you are trying to install "symfony dev-master"?!

**Comment by mdogancay** [comment]
> looking at the output, it more looks like you are trying to install "symfony dev-master"?!

composer req azine/email-bundle dev-master

**Comment by mdogancay** [comment]
composer req azine/email-bundle
Using version ^3.0 for azine/email-bundle
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Restricting packages listed in "symfony/symfony" to "4.2.*"
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for azine/email-bundle ^3.0 -> satisfiable by azine/email-bundle[3.0.0].
    - azine/email-bundle 3.0.0 requires symfony/framework-bundle ~2.7,>=2.7 -> no matching package found.

I tried everything it can't be installed on Symfony 4.2.4

**Comment by azine** [comment]
Did you try with Symfony "4.x" and azine-email "dev-master"?

**Comment by azine** [comment]
Can you post your composer.json here?

**Comment by mdogancay** [comment]
> Can you post your composer.json here?

Not necessary(I am using latest symfony);
Start a new project with symfony/website-skeleton template:
composer create-project symfony/website-skeleton testproject

try to install your bundle with:
composer require azine/email-bundle
or
composer require azine/email-bundle dev-master
or
composer require azine/email-bundle:dev-master
It will not install.

**Comment by mdogancay** [comment]
> Can you post your composer.json here?
here is the test project composer.json:
`{
    "type": "project",
    "license": "proprietary",
    "require": {
        "php": "^7.1.3",
        "ext-ctype": "*",
        "ext-iconv": "*",
        "sensio/framework-extra-bundle": "^5.1",
        "symfony/asset": "4.2.*",
        "symfony/console": "4.2.*",
        "symfony/dotenv": "4.2.*",
        "symfony/expression-language": "4.2.*",
        "symfony/flex": "^1.1",
        "symfony/form": "4.2.*",
        "symfony/framework-bundle": "4.2.*",
        "symfony/monolog-bundle": "^3.1",
        "symfony/orm-pack": "*",
        "symfony/process": "4.2.*",
        "symfony/security-bundle": "4.2.*",
        "symfony/serializer-pack": "*",
        "symfony/swiftmailer-bundle": "^3.1",
        "symfony/translation": "4.2.*",
        "symfony/twig-bundle": "4.2.*",
        "symfony/validator": "4.2.*",
        "symfony/web-link": "4.2.*",
        "symfony/yaml": "4.2.*"
    },
    "require-dev": {
        "symfony/debug-pack": "*",
        "symfony/maker-bundle": "^1.0",
        "symfony/profiler-pack": "*",
        "symfony/test-pack": "*",
        "symfony/web-server-bundle": "4.2.*"
    },
    "config": {
        "preferred-install": {
            "*": "dist"
        },
        "sort-packages": true
    },
    "autoload": {
        "psr-4": {
            "App\\": "src/"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "App\\Tests\\": "tests/"
        }
    },
    "replace": {
        "paragonie/random_compat": "2.*",
        "symfony/polyfill-ctype": "*",
        "symfony/polyfill-iconv": "*",
        "symfony/polyfill-php71": "*",
        "symfony/polyfill-php70": "*",
        "symfony/polyfill-php56": "*"
    },
    "scripts": {
        "auto-scripts": {
            "cache:clear": "symfony-cmd",
            "assets:install %PUBLIC_DIR%": "symfony-cmd"
        },
        "post-install-cmd": [
            "@auto-scripts"
        ],
        "post-update-cmd": [
            "@auto-scripts"
        ]
    },
    "conflict": {
        "symfony/symfony": "*"
    },
    "extra": {
        "symfony": {
            "allow-contrib": false,
            "require": "4.2.*"
        }
    }
}`

**Comment by azine** [comment]
hmmmm... that's strange. installing the following composer.json works fine.

```yml
{
    "require": {
        "symfony/symfony": "^4.2",
        "azine/email-bundle": "dev-master"
    }
}
```

I get the following output by `composer install`:

```bash
$ php composer.phar install
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 38 installs, 0 updates, 0 removals
  - Installing symfony/polyfill-ctype (v1.10.0): Loading from cache
  - Installing paragonie/random_compat (v2.0.18): Loading from cache
  - Installing ramsey/uuid (3.8.0): Loading from cache
  - Installing symfony/polyfill-mbstring (v1.10.0): Loading from cache
  - Installing twig/twig (v2.7.2): Loading from cache
  - Installing symfony/polyfill-php72 (v1.10.0): Loading from cache
  - Installing symfony/symfony (v4.2.4): Downloading (100%)         
  - Installing symfony/polyfill-intl-icu (v1.10.0): Loading from cache
  - Installing symfony/contracts (v1.0.2): Downloading (100%)         
  - Installing psr/simple-cache (1.0.1): Downloading (100%)         
  - Installing psr/log (1.1.0): Loading from cache
  - Installing psr/link (1.0.0): Downloading (100%)         
  - Installing psr/container (1.0.0): Downloading (100%)         
  - Installing psr/cache (1.0.1): Downloading (100%)         
  - Installing fig/link-util (1.0.0): Downloading (100%)         
  - Installing doctrine/lexer (v1.0.1): Loading from cache
  - Installing doctrine/annotations (v1.6.0): Loading from cache
  - Installing doctrine/reflection (v1.0.0): Downloading (100%)         
  - Installing doctrine/event-manager (v1.0.0): Downloading (100%)         
  - Installing doctrine/collections (v1.5.0): Loading from cache
  - Installing doctrine/cache (v1.8.0): Loading from cache
  - Installing doctrine/persistence (v1.1.0): Downloading (100%)         
  - Installing monolog/monolog (1.24.0): Loading from cache
  - Installing friendsofsymfony/user-bundle (v2.1.2): Loading from cache
  - Installing azine/emailupdateconfirmation-bundle (1.0.1): Downloading (100%)         
  - Installing symfony/polyfill-intl-idn (v1.10.0): Downloading (100%)         
  - Installing symfony/polyfill-iconv (v1.10.0): Downloading (100%)         
  - Installing egulias/email-validator (2.1.7): Downloading (100%)         
  - Installing swiftmailer/swiftmailer (v6.2.0): Downloading (100%)         
  - Installing doctrine/instantiator (1.1.0): Loading from cache
  - Installing doctrine/dbal (v2.9.2): Downloading (100%)         
  - Installing doctrine/inflector (v1.3.0): Loading from cache
  - Installing doctrine/common (v2.10.0): Downloading (100%)         
  - Installing doctrine/orm (v2.6.3): Downloading (100%)         
  - Installing knplabs/knp-components (v1.3.10): Downloading (100%)         
  - Installing knplabs/knp-paginator-bundle (v2.8.0): Downloading (100%)         
  - Installing twig/extensions (v1.5.4): Loading from cache
  - Installing azine/email-bundle (dev-master 47695d5): Cloning 47695d5a4c from cache
paragonie/random_compat suggests installing ext-libsodium (Provides a modern crypto API that can be used to generate random bytes.)
ramsey/uuid suggests installing ircmaxell/random-lib (Provides RandomLib for use with the RandomLibAdapter)
ramsey/uuid suggests installing ext-libsodium (Provides the PECL libsodium extension for use with the SodiumRandomGenerator)
ramsey/uuid suggests installing ext-uuid (Provides the PECL UUID extension for use with the PeclUuidTimeGenerator and PeclUuidRandomGenerator)
ramsey/uuid suggests installing moontoast/math (Provides support for converting UUID to 128-bit integer (in string form).)
ramsey/uuid suggests installing ramsey/uuid-doctrine (Allows the use of Ramsey\Uuid\Uuid as Doctrine field type.)
ramsey/uuid suggests installing ramsey/uuid-console (A console application for generating UUIDs with ramsey/uuid)
symfony/contracts suggests installing symfony/cache-contracts-implementation
symfony/contracts suggests installing symfony/service-contracts-implementation
symfony/contracts suggests installing symfony/translation-contracts-implementation
doctrine/cache suggests installing alcaeus/mongo-php-adapter (Required to use legacy MongoDB driver)
monolog/monolog suggests installing aws/aws-sdk-php (Allow sending log messages to AWS services like DynamoDB)
monolog/monolog suggests installing doctrine/couchdb (Allow sending log messages to a CouchDB server)
monolog/monolog suggests installing ext-amqp (Allow sending log messages to an AMQP server (1.0+ required))
monolog/monolog suggests installing ext-mongo (Allow sending log messages to a MongoDB server)
monolog/monolog suggests installing graylog2/gelf-php (Allow sending log messages to a GrayLog2 server)
monolog/monolog suggests installing mongodb/mongodb (Allow sending log messages to a MongoDB server via PHP Driver)
monolog/monolog suggests installing php-amqplib/php-amqplib (Allow sending log messages to an AMQP server using php-amqplib)
monolog/monolog suggests installing php-console/php-console (Allow sending log messages to Google Chrome)
monolog/monolog suggests installing rollbar/rollbar (Allow sending log messages to Rollbar)
monolog/monolog suggests installing ruflin/elastica (Allow sending log messages to an Elastic Search server)
monolog/monolog suggests installing sentry/sentry (Allow sending log messages to a Sentry server)
swiftmailer/swiftmailer suggests installing true/punycode (Needed to support internationalized email addresses, if ext-intl is not installed)
Writing lock file
Generating autoload files
```

Installing `azine/email-bundle` version 3.0 does indeed not work with Symfony 4.x. I'll have to tag a versoion 4.0 of the `azine/email-bundle` so installation works without relying on `dev-master`.

**Comment by azine** [comment]
I have just tagged v4.0.0 of the email-bundle, now `composer req azine/email-bundle` works too and installs `^4.0` on my machine.

**Comment by mdogancay** [comment]
> I have just tagged v4.0.0 of the email-bundle, now `composer req azine/email-bundle` works too and installs `^4.0` on my machine.
> 
> Please close the issue if you problem is solved with this.

`xxxuser@alphai5:~/projects/test-projects/emailtest$ composer req azine/email-bundle
Using version ^4.0 for azine/email-bundle
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Restricting packages listed in "symfony/symfony" to "4.2.*"
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for __root__ No version set (parsed as 1.0.0) -> satisfiable by __root__[No version set (parsed as 1.0.0)].
    - azine/email-bundle 4.0.0 requires azine/emailupdateconfirmation-bundle ~1.0 -> satisfiable by azine/emailupdateconfirmation-bundle[1.0.0, 1.0.1].
    - azine/emailupdateconfirmation-bundle 1.0.0 requires symfony/symfony ~2.7|^3.0|^4.0 -> satisfiable by symfony/symfony[v4.2.0, v4.2.1, v4.2.2, v4.2.3, v4.2.4].
    - azine/emailupdateconfirmation-bundle 1.0.1 requires symfony/symfony ~2.7|^3.0|^4.0 -> satisfiable by symfony/symfony[v4.2.0, v4.2.1, v4.2.2, v4.2.3, v4.2.4].
    - symfony/symfony v4.2.0 conflicts with __root__[No version set (parsed as 1.0.0)].
    - symfony/symfony v4.2.1 conflicts with __root__[No version set (parsed as 1.0.0)].
    - symfony/symfony v4.2.2 conflicts with __root__[No version set (parsed as 1.0.0)].
    - symfony/symfony v4.2.3 conflicts with __root__[No version set (parsed as 1.0.0)].
    - symfony/symfony v4.2.4 conflicts with __root__[No version set (parsed as 1.0.0)].
    - Installation request for azine/email-bundle ^4.0 -> satisfiable by azine/email-bundle[4.0.0].


Installation failed, reverting ./composer.json to its original content.
`

Not working on fresh install symfony 4.2.4 website skeleton.

**Comment by mdogancay** [comment]
> I have just tagged v4.0.0 of the email-bundle, now `composer req azine/email-bundle` works too and installs `^4.0` on my machine.
> 
> Please close the issue if you problem is solved with this.
Not installing on fresh created "symfony/website-skeleton" or my project either.

`xxuser@alphai5:~/projects/test-projects/emailtest$ composer req azine/email-bundle
Using version ^4.0 for azine/email-bundle
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Restricting packages listed in "symfony/symfony" to "4.2.*"
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for __root__ No version set (parsed as 1.0.0) -> satisfiable by __root__[No version set (parsed as 1.0.0)].
    - azine/email-bundle 4.0.0 requires azine/emailupdateconfirmation-bundle ~1.0 -> satisfiable by azine/emailupdateconfirmation-bundle[1.0.0, 1.0.1].
    - azine/emailupdateconfirmation-bundle 1.0.0 requires symfony/symfony ~2.7|^3.0|^4.0 -> satisfiable by symfony/symfony[v4.2.0, v4.2.1, v4.2.2, v4.2.3, v4.2.4].
    - azine/emailupdateconfirmation-bundle 1.0.1 requires symfony/symfony ~2.7|^3.0|^4.0 -> satisfiable by symfony/symfony[v4.2.0, v4.2.1, v4.2.2, v4.2.3, v4.2.4].
    - symfony/symfony v4.2.0 conflicts with __root__[No version set (parsed as 1.0.0)].
    - symfony/symfony v4.2.1 conflicts with __root__[No version set (parsed as 1.0.0)].
    - symfony/symfony v4.2.2 conflicts with __root__[No version set (parsed as 1.0.0)].
    - symfony/symfony v4.2.3 conflicts with __root__[No version set (parsed as 1.0.0)].
    - symfony/symfony v4.2.4 conflicts with __root__[No version set (parsed as 1.0.0)].
    - Installation request for azine/email-bundle ^4.0 -> satisfiable by azine/email-bundle[4.0.0].


Installation failed, reverting ./composer.json to its original content.
`

**Comment by azine** [comment]
I found the issue. Symfony-Skeleton is a `flex` application, which explicitly conflicts with `symfony/symfony`. But `azine/email-bundle` depends on `azine/emailupdateconfirmation-bundle`, which in turn depends on classes from the `Symfony\Component\` namespace. Currently `azine/emailupdateconfirmation-bundle` includes these classes via requiring `symfony/symfony`.

So the issue is basically that this bundle does not (yet) support `flex` (because `azine/emailupdateconfirmation-bundle` does not (yet) support `flex`).

**Comment by mdogancay** [comment]
> I found the issue. Symfony-Skeleton is a `flex` application, which explicitly conflicts with `symfony/symfony`. But `azine/email-bundle` depends on `azine/emailupdateconfirmation-bundle`, which in turn depends on classes from the `Symfony\Component\` namespace. Currently `azine/emailupdateconfirmation-bundle` includes these classes via requiring `symfony/symfony`.
> 
> So the issue is basically that this bundle does not (yet) support `flex` (because `azine/emailupdateconfirmation-bundle` does not (yet) support `flex`).

Super, we know the broblem, how we can fix it? How can I install your bundle on my project? Any idea?

**Comment by mdogancay** [comment]
> Someone has to update the composer.json of the azine/email-bundle and the azine/emailupdateconfirmation-bundle (and in the worst case also some of the code) with a set of dependencies that complies with Symfony flex => not depend on `symfony/symfony`.
> 
> OR
> 
> You change your project to not be a symfony flex application (and not depend on `symfony/website-skeleton`
> 
> PRs are welcome. Unfortunately it will be a while before one of us will come around to look at this.
> 
> See: [azine/AzineEmailUpdateConfirmationBundle#7](https://github.com/azine/AzineEmailUpdateConfirmationBundle/issues/7)

Ok. thank you for your time and help. I really appreciated

**Comment by azine** [comment]
Hi @mdogancay,

I had some time spare today. Give it a try again with the latest "azine/email-bundle" => 4.0.1

**Comment by mdogancay** [comment]
Hi @azine,
I tried on fresh install symfony 4.2.4 (`composer create-project symfony/website-skeleton`)
It was installed with an error here is the output :
```
$ composer req azine/email-bundle
Using version ^4.0 for azine/email-bundle
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Restricting packages listed in "symfony/symfony" to "4.2.*"

Prefetching 7 packages 🎶 💨
  - Downloading (100%)

Package operations: 9 installs, 0 updates, 0 removals
  - Installing symfony/lock (v4.2.4): Loading from cache
  - Installing symfony/templating (v4.2.4): Loading from cache
  - Installing friendsofsymfony/user-bundle (v2.1.2): Loading from cache
  - Installing azine/emailupdateconfirmation-bundle (1.1.0): Loading from cache
  - Installing knplabs/knp-components (v1.3.10): Loading from cache
  - Installing knplabs/knp-paginator-bundle (v2.8.0): Loading from cache
  - Installing ramsey/uuid (3.8.0): Loading from cache
  - Installing twig/extensions (v1.5.4): Loading from cache
  - Installing azine/email-bundle (4.0.1): Loading from cache
Writing lock file
Generating autoload files
ocramius/package-versions:  Generating version class...
ocramius/package-versions: ...done generating version class
Symfony operations: 5 recipes (9461d46a26d56aa37a1cdc3a230d2526)
  - Configuring friendsofsymfony/user-bundle (>=v2.1.2): From auto-generated recipe
  - Configuring azine/emailupdateconfirmation-bundle (>=1.1.0): From auto-generated recipe
  - Configuring knplabs/knp-paginator-bundle (>=v2.8.0): From auto-generated recipe
  - Configuring twig/extensions (>=1.0): From github.com/symfony/recipes:master
  - Configuring azine/email-bundle (>=4.0.1): From auto-generated recipe
Executing script cache:clear [KO]
 [KO]
Script cache:clear returned with error code 1
!!  
!!  In ArrayNode.php line 228:
!!                                                                       
!!    The child node "db_driver" at path "fos_user" must be configured.  
!!                                                                       
!!  
!!  
Script @auto-scripts was called via post-update-cmd
```

**Comment by mdogancay** [comment]
I tested the functionality : 
```
projectDir$ bin/console emails:sendNewsletter
Thu, 21 Mar 2019 13:32:16 +0300 : starting to send newsletter emails.

In Container.php line 263:
                                                                                                                                                                                                       
  The "azine_email_notifier_service" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dep  
  endency injection instead.
```
and its not working. I tried to inject service to the command but it seems not working.
I thing, It's needs more and more fix for the work in latest symfony (4.2.4)
thanks anyway.

**Comment by azine** [comment]
Happy to see that the installation works. 

Regarding the last error message you receive: to be able to send a newsletter you have to implement your version of this service. See https://github.com/azine/email-bundle#customise-the-content-and-subjects-of-your-emails

I am not sure if the "ExampleService" provided in this Bundle works out of the box. Your error message suggests it doesn't. But either way, as noted in the README.md, you will have to implement a few services and configure them to be used.
